### PR TITLE
Fix save and load for TPM2B_PRIVATE

### PR DESCRIPTION
**Pull request by williamcroberts**
Signed-off-by: William Roberts <william.c.roberts@intel.com>

### CODE REVIEW

**Comment by martinezjavier** [comment]
@williamcroberts how did you test it? doesn't compile for me. The fix is trivial though:

```
diff --git a/tools/tpm2_create.c b/tools/tpm2_create.c
index d896e66cd6a2..41d7b428deed 100644
--- a/tools/tpm2_create.c
+++ b/tools/tpm2_create.c
@@ -220,7 +220,7 @@ int create(TSS2_SYS_CONTEXT *sapi_context)
     }
 
     if (ctx.flags.O) {
-        bool res = files_save_private(ctx.opr_path, &outPrivate);
+        bool res = files_save_private(&outPrivate, ctx.opr_path);
         if (!res) {
             return -4;
         }
```

It does fix the issue though since after storing the size field, I can load the key and unseal using master:

```
$ export TPM2TOOLS_TCTI=device:/dev/tpmrm0

$ tpm2_load -c primary.context  -u obj.pub -r obj.priv -n obj.name -C load.context                                                                                                
handle: 0x80fffffe
$ tpm2_unseal -c load.context -L sha1:0,1,2,3,4 -F pcr.bin
secret

```

**Comment by martinezjavier** [comment]
@williamcroberts pleaes add a Fixes: #976 to your commit message.

File: tools/tpm2_create.c [hunk 1]
```diff
@@ -220,7 +220,7 @@ int create(TSS2_SYS_CONTEXT *sapi_context)
     }
 
     if (ctx.flags.O) {
-        bool res = files_save_bytes_to_file(ctx.opr_path, outPrivate.t.buffer, outPrivate.t.size);
+        bool res = files_save_private(ctx.opr_path, &outPrivate);
```
Comment (martinezjavier) [inline tools/tpm2_create.c]: I would just include this part and the others that fix the issue and avoid the other bits that deal with the wrong size. This was clearly a bug so I would just fix it. Although the others changes cause no harm so up to you.

**Comment by williamcroberts** [comment]
<snip>
> 
> > @@ -220,7 +220,7 @@ int create(TSS2_SYS_CONTEXT *sapi_context)
>      }
> 
>      if (ctx.flags.O) {
> -        bool res = files_save_bytes_to_file(ctx.opr_path, outPrivate.t.buffer,
> outPrivate.t.size);
> +        bool res = files_save_private(ctx.opr_path, &outPrivate);
> 
> @williamcroberts <https://github.com/williamcroberts>  right, sorry for
> explaining so badly. What I tried to say is that this line is what fixes the
> serialization issue since by also storing the size field of the structure.
> 
> And most of the changes in lib/files.c are to cope with loading an object of wrong
> size, and I said that I wouldn't include that.
> 

Oh, so what about a file saved via the broken version, the LOAD won't work, I thought
You didn't want to break that?

**Comment by martinezjavier** [comment]
> Oh, so what about a file saved via the broken version, the LOAD won't work, I thought
> You didn't want to break that?

@williamcroberts  well, I thought that we shouldn't break backward compatibility on purpose for the saved files between major releases, but I think it's different if the store of the private portion was already broken. In other words, I think it's different fixing a serialization bug than changing how things are stored between releases.

Also, AFAICT no major distro has already shipped tpm2-tools 3.X on a release. Fedora 28 and openSUSE Leap 15.0 will be released soon with tpm2-tools 3.X and Ubuntu Bionic still ships the 2.X version. So I think it's still safe to just fix this bug without the need to do a conversion on load.

I remember we found others serialization bugs in the past and didn't add code to cope with that, but I don't have a strong position on this. As I said, the code causes no harm other than maybe having code that may not be used.

**Comment by williamcroberts** [comment]
<snip

> 
> Also, AFAICT no major distro has already shipped tpm2-tools 3.X on a release.
> Fedora 28 and openSUSE Leap 15.0 will be released soon with tpm2-tools 3.X and
> Ubuntu Bionic still ships the 2.X version. So I think it's still safe to just fix this bug
> without the need to do a conversion on load.

SGTM

> 
> I remember we found others serialization bugs in the past and didn't add code to
> cope with that, but I don't have a strong position on this. As I said, the code
> causes no harm other than maybe having code that may not be used.
> 

Speaking of that, I think the next release should really be 3.X with these fixes, and
Then a 3.1 working on the newer version of the TSS code.

Do you want to spin these and become the stable branch guy?

**Comment by martinezjavier** [comment]
> Speaking of that, I think the next release should really be 3.X with these fixes, and
> Then a 3.1 working on the newer version of the TSS code.

Not sure I understood correctly, I think we agree that a 3.0.4 release will be 3.0.3 + all the fixes in the 3.X branc (plus your TPM2B_PRIVATE save / load fix). 

And 3.1 will be a release from the master branch once tpm2-tss has a release?

> Do you want to spin these and become the stable branch guy?

Do you mean making a 3.0.4 tag, creating and signing the tarballs, etc? Sure thing, I can do the 3.0.4 on Monday. I can follow the RELEASE.md and ask you for assistance if something isn't clear.

**Comment by williamcroberts** [comment]
> -----Original Message-----
> From: Javier Martinez Canillas [mailto:notifications@github.com]
> Sent: Friday, April 27, 2018 2:50 PM
> To: tpm2-software/tpm2-tools <tpm2-tools@noreply.github.com>
> Cc: Roberts, William C <william.c.roberts@intel.com>; Mention
> <mention@noreply.github.com>
> Subject: Re: [tpm2-software/tpm2-tools] Fix save and load for TPM2B_PRIVATE
> (#1014)
> 
> 	Speaking of that, I think the next release should really be 3.X with these
> fixes, and
> 	Then a 3.1 working on the newer version of the TSS code.
> 
> Not sure I understood correctly, I think we agree that a 3.0.4 release will be 3.0.3
> + all the fixes in the 3.X branc (plus your TPM2B_PRIVATE save / load fix).

3.0.4 will have all the patches on 3.X

The next 3.X release IMO, will be 3.1.0 re-worked ontop of the newest TSS version (master)
currently, but that will be getting RCs at some point.

> 
> And 3.1 will be a release from the master branch once tpm2-tss has a release?
> 
> 	Do you want to spin these and become the stable branch guy?
> 
> Do you mean making a 3.0.4 tag, creating and signing the tarballs, etc? Sure thing,
> I can do the 3.0.4 on Monday. I can follow the RELEASE.md and ask you for
> assistance if something isn't clear.

Yep.

**Comment by martinezjavier** [comment]
I'm closing this pull request since it has been superseded by #1018 "[v2] tools: Fix save and load for TPM2B_PRIVATE".

**Pull request closed without merging**